modules/diskq/tests/test_diskq_truncate.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2018 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "logqueue.h"
+#include "logqueue-fifo.h"
+#include "logqueue-disk.h"
+#include "logqueue-disk-reliable.h"
+#include "logqueue-disk-non-reliable.h"
+#include "diskq.h"
+#include "qdisk.c"
+#include "apphook.h"
+
+#include "queue_utils_lib.h"
+#include "test_diskq_tools.h"
+#include "testutils.h"
+
+#include <stdlib.h>
+#include <sys/stat.h>
+
+
+MsgFormatOptions parse_options;
+
+static LogQueue *
+_get_diskqueue(gchar *filename, DiskQueueOptions *options)
+{
+  LogQueue *q = log_queue_disk_non_reliable_new(options, NULL);
+  log_queue_set_use_backlog(q, FALSE);
+  log_queue_disk_load_queue(q, filename);
+  return q;
+}
+
+static LogQueue *
+_create_diskqueue(gchar *filename, DiskQueueOptions *options)
+{
+  LogQueue *q;
+  unlink(filename);
+
+  _construct_options(options, 1100000, 0, FALSE);
+  options->qout_size = 40;
+  q = _get_diskqueue(filename, options);
+  return q;
+}
+
+static void
+_save_diskqueue(LogQueue *q)
+{
+  gboolean persistent;
+  log_queue_disk_save_queue(q, &persistent);
+  log_queue_unref(q);
+}
+
+typedef struct
+{
+  gint number_of_msgs_to_push;
+  gint number_of_msgs_to_pop;
+  gboolean should_file_be_empty_after_truncate;
+} truncate_test_params;
+
+static void
+_test_diskq_truncate(truncate_test_params params)
+{
+  LogQueue *q;
```

Comment:
As we discussed, using `file_size` is not optimal:
 if disk_queue truncate logic is broken, but `file_size` updated accordingly and we just ask the file size with `stat` then we wouldn't see the incorrent truncate.
A better expected value needed when testing truncate.

Also, it's not enough to check only if the file is not empty.
We agreed that one way to do this is to calculate the expected size after truncate:
- serialize 1 input message as it happens in the disk-queue
- multiply with the number of expected messages
- also take into consideration the header size (QDISK_RESERVED_SPACE)
This vale can be used against a `stat()` result as expected value.

The only drawback is, that we don't use an API of the disk-queue, but reimplement something that syslog-ng does, hence it can be easily broke as well.
Hopefully not silently.